src/Storages/ExternalDataSourceConfiguration.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/Context.h>
+#include <Poco/Util/AbstractConfiguration.h>
+
+
+namespace DB
+{
+
+struct ExternalDataSourceConfiguration
+{
+    String host;
+    UInt16 port = 0;
+    String username;
+    String password;
+    String database;
+    String table;
+    String schema;
+
+    std::vector<std::pair<String, UInt16>> addresses; /// Failover replicas.
+
+    String toString() const;
+};
+
+using ExternalDataSourceConfigurationPtr = std::shared_ptr<ExternalDataSourceConfiguration>;
+
+
+struct StoragePostgreSQLConfiguration : ExternalDataSourceConfiguration
+{
+    explicit StoragePostgreSQLConfiguration(const ExternalDataSourceConfiguration & common_configuration)
+        : ExternalDataSourceConfiguration(common_configuration) {}
+
+    String on_conflict;
+};
+
+
+struct StorageMySQLConfiguration : ExternalDataSourceConfiguration
+{
+    explicit StorageMySQLConfiguration(const ExternalDataSourceConfiguration & common_configuration)
+        : ExternalDataSourceConfiguration(common_configuration) {}
+
+    bool replace_query = false;
+    String on_duplicate_clause;
+};
+
+struct StorageMongoDBConfiguration : ExternalDataSourceConfiguration
+{
+    explicit StorageMongoDBConfiguration(const ExternalDataSourceConfiguration & common_configuration)
+        : ExternalDataSourceConfiguration(common_configuration) {}
+
+    String collection;
+    String options;
+};
+
+
+using EngineArgs = std::vector<std::pair<String, DB::Field>>;
+
+/* If storage engine's configuration was define via named_collections,
+ * return all options in ExternalDataSource::Configuration struct.
+ *
+ * Also check if engine arguments have key-value defined configuration options:
+ * ENGINE = PostgreSQL(postgresql_configuration, database = 'postgres_database');
+ * In this case they will override values defined in config.
+ *
+ * If there are key-value arguments apart from common: `host`, `port`, `username`, `password`, `database`,
+ * i.e. storage-specific arguments, then return them back in a set: ExternalDataSource::EngineArgs.
+ */
+std::tuple<ExternalDataSourceConfiguration, EngineArgs, bool>
```

Comment:
No, actually I think this way will be better.
Now I can have a result of this method as `auto [configuration, args, is_named_collection] = func()`.
But if I use std::optional like you way it will be not as good looking.